llpc/tool/llpcComputePipelineBuilder.cpp

```diff
@@ -0,0 +1,159 @@
+// @returns : Pipeline binary data on success, `llvm::ResultError` on failure.
+Expected<BinaryData> ComputePipelineBuilder::buildComputePipeline() {
+  CompileInfo &compileInfo = getCompileInfo();
+  assert(compileInfo.shaderModuleDatas.size() == 1);
+
+  StandaloneCompiler::ShaderModuleData &moduleData = compileInfo.shaderModuleDatas[0];
+  assert(moduleData.shaderStage == ShaderStageCompute);
+
+  ComputePipelineBuildInfo *pipelineInfo = &compileInfo.compPipelineInfo;
+  ComputePipelineBuildOut *pipelineOut = &compileInfo.compPipelineOut;
+
+  PipelineShaderInfo *shaderInfo = &pipelineInfo->cs;
+  const ShaderModuleBuildOut *shaderOut = &moduleData.shaderOut;
+
+  // If entry target is not specified, use the one from command line option.
+  if (!shaderInfo->pEntryTarget)
+    shaderInfo->pEntryTarget = moduleData.entryPoint.c_str();
+
+  shaderInfo->entryStage = ShaderStageCompute;
+  shaderInfo->pModuleData = shaderOut->pModuleData;
+
+  // If not compiling from pipeline, lay out user data now.
+  if (compileInfo.doAutoLayout) {
+    ResourceMappingNodeMap nodeSets;
+    unsigned pushConstSize = 0;
+    doAutoLayoutDesc(ShaderStageCompute, moduleData.spirvBin, nullptr, shaderInfo, nodeSets, pushConstSize,
+                     /*autoLayoutDesc =*/compileInfo.autoLayoutDesc);
+
+    buildTopLevelMapping(ShaderStageComputeBit, nodeSets, pushConstSize, &pipelineInfo->resourceMapping,
+                         compileInfo.autoLayoutDesc);
+  }
+
+  pipelineInfo->pInstance = nullptr; // Placeholder, unused.
+  pipelineInfo->pUserData = &compileInfo.pipelineBuf;
+  pipelineInfo->pfnOutputAlloc = allocateBuffer;
+  pipelineInfo->unlinked = compileInfo.unlinked;
+  pipelineInfo->options.robustBufferAccess = compileInfo.robustBufferAccess;
+  pipelineInfo->options.enableRelocatableShaderElf = compileInfo.relocatableShaderElf;
+  pipelineInfo->options.scalarBlockLayout = compileInfo.scalarBlockLayout;
+  pipelineInfo->options.enableScratchAccessBoundsChecks = compileInfo.scratchAccessBoundsChecks;
+
+  PipelineBuildInfo localPipelineInfo = {};
+  localPipelineInfo.pComputeInfo = pipelineInfo;
+  void *pipelineDumpHandle = runPreBuildActions(localPipelineInfo);
+  auto onExit = make_scope_exit([&] { runPostBuildActions(pipelineDumpHandle, {pipelineOut->pipelineBin}); });
+
+  ICompiler &compiler = getCompiler();
+  Result result = compiler.BuildComputePipeline(pipelineInfo, pipelineOut, pipelineDumpHandle);
+  if (result != Result::Success)
+    return createResultError(result, "Compute pipeline compilation failed");
+
+  return pipelineOut->pipelineBin;
+}
+
+// =====================================================================================================================
+// Calculates the pipeline hash.
+//
+// @param buildInfo : Pipeline build info.
+// @returns : Calculated pipeline hash.
+uint64_t ComputePipelineBuilder::getPipelineHash(Vkgc::PipelineBuildInfo buildInfo) {
+  return IPipelineDumper::GetPipelineHash(buildInfo.pGraphicsInfo);
```

Comment:
should this be `pComputeInfo`?

Reply:
🤦‍♂️

Reply:
I'm surprised none of the tests caught this. I think we only check that pipeline dumps are emitted, but not that we can round trip.

Reply:
The code coverage report does indeed show that line is not exercised.

Reply:
I opened a PR to add some basic tests: https://github.com/GPUOpen-Drivers/llpc/pull/1630

Reply:
I also opened a PR with pipeline info tests: https://github.com/GPUOpen-Drivers/llpc/pull/1633